### DISCUSSION TITLE
AssertionError response when attempting to create cable

### DISCUSSION THREAD

**Discussion by mblue-fasttrack** [netbox-community/netbox#12076] [Help Wanted!] [answered]
I receive an assertion error whenever I attempt to create a cable connection between 2 specific front ports and another connection. I am able to connect other pairs of front ports via cable. This issue seems to occur on 2 specific front ports on a single device.
![2023-03-27 21_42_33-Add a new cable _ NetBox](https://user-images.githubusercontent.com/105229887/228123858-e02f9459-aa14-48d4-861f-776f082c9584.png)
![2023-03-27 21_43_05-Server Error](https://user-images.githubusercontent.com/105229887/228123867-2a1c1758-fcd1-476f-a2e7-e380f88e963d.png)

**Comment by candlerb** [comment]
Sounds like your database has got into an inconsistent state (which can happen, since Netbox/Django model validation is subject to various race conditions)

If you can get a backtrace from your error, that will help. You can get this by setting EMAIL and ADMINS in configuration.py, and then reproducing the error - a full report including the backtrace should be emailed to you. Or you can set DEBUG=True and get the backtrace in the web interface.

**Reply by mblue-fasttrack** [reply]
Trackback info below:

Environment:


Request Method: POST
Request URL: https://[redacted]/dcim/cables/add/?a_terminations_type=dcim.frontport&a_terminations=1597&b_terminations_type=dcim.interface&termination_b_site=169&termination_b_rack=17&return_url=/dcim/devices/102/front-ports/

Django Version: 4.1.5
Python Version: 3.8.10
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'django.contrib.humanize',
 'corsheaders',
 'debug_toolbar',
 'graphiql_debug_toolbar',
 'django_filters',
 'django_tables2',
 'django_prometheus',
 'graphene_django',
 'mptt',
 'rest_framework',
 'social_django',
 'taggit',
 'timezone_field',
 'circuits',
 'dcim',
 'ipam',
 'extras',
 'tenancy',
 'users',
 'utilities',
 'virtualization',
 'wireless',
 'django_rq',
 'drf_yasg']
Installed Middleware:
['graphiql_debug_toolbar.middleware.DebugToolbarMiddleware',
 'django_prometheus.middleware.PrometheusBeforeMiddleware',
 'corsheaders.middleware.CorsMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'django.middleware.security.SecurityMiddleware',
 'netbox.middleware.ExceptionHandlingMiddleware',
 'netbox.middleware.RemoteUserMiddleware',
 'netbox.middleware.LoginRequiredMiddleware',
 'netbox.middleware.DynamicConfigMiddleware',
 'netbox.middleware.APIVersionMiddleware',
 'netbox.middleware.ObjectChangeMiddleware',
 'django_prometheus.middleware.PrometheusAfterMiddleware']



Traceback (most recent call last):
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/generic/base.py", line 103, in view
    return self.dispatch(request, *args, **kwargs)
  File "/opt/netbox/netbox/dcim/views.py", line 3197, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/opt/netbox/netbox/netbox/views/generic/object_views.py", line 167, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/opt/netbox/netbox/netbox/views/generic/base.py", line 26, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/opt/netbox/netbox/utilities/views.py", line 99, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/generic/base.py", line 142, in dispatch
    return handler(request, *args, **kwargs)
  File "/opt/netbox/netbox/netbox/views/generic/object_views.py", line 255, in post
    obj = form.save()
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/forms/models.py", line 548, in save
    self.instance.save()
  File "/opt/netbox/netbox/dcim/models/cables.py", line 217, in save
    trace_paths.send(Cable, instance=self, created=_created)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 176, in send
    return [
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/dispatch/dispatcher.py", line 177, in <listcomp>
    (receiver, receiver(signal=self, sender=sender, **named))
  File "/opt/netbox/netbox/dcim/signals.py", line 97, in update_connected_endpoints
    create_cablepath(nodes)
  File "/opt/netbox/netbox/dcim/utils.py", line 43, in create_cablepath
    cp = CablePath.from_origin(terminations)
  File "/opt/netbox/netbox/dcim/models/cables.py", line 553, in from_origin
    assert all(rp.positions == 1 for rp in rear_ports)

Exception Type: AssertionError at /dcim/cables/add/
Exception Value:

**Comment by candlerb** [comment, marked as answer]
> assert all(rp.positions == 1 for rp in rear_ports)

OK, here's some context for that code: it's trying to trace a cable path.

```
            # Step 6: Determine the "next hop" terminations, if applicable
            if not remote_terminations:
                break

            if isinstance(remote_terminations[0], FrontPort):
                # Follow FrontPorts to their corresponding RearPorts
                rear_ports = RearPort.objects.filter(
                    pk__in=[t.rear_port_id for t in remote_terminations]
                )
                if len(rear_ports) > 1:
                    assert all(rp.positions == 1 for rp in rear_ports)     <<<< EXCEPTION HERE
                elif rear_ports[0].positions > 1:
                    position_stack.append([fp.rear_port_position for fp in remote_terminations])

                terminations = rear_ports
```

It seems that if there are multiple rearports linked to the same cable, then each rearport is expected to have a single position only.  (A rearport with multiple "positions" is one that links to multiple frontports)

Try tracing from UPG Rx and UPG Tx, using the automatic cable tracing and/or manually following cables, and look at each rearport on the path, and see which ones have been defined to have multiple positions, and if that's what you expect.

You may be able to find what's different about these two particular frontports that causes the crash.  For example, even if each frontport is only linked to a single rearport, the rearport might have been *defined* to have more than 1 position.

**Reply by mblue-fasttrack** [reply]
It seems like this error only occurs if multiple front ports attached to different, multi-position rear ports are attached to a single interface. Attaching the front ports to other front ports works without issue.

The front ports causing the error are supposed to be attached to different multi-position rear ports, meaning this is expected. We are modeling WDM filters with 2 multi-position rear ports (TX and RX) each linked to their respective TX and RX wavelengths in the front. Is this not a supported method of modeling these in Netbox? Is Netbox able to support an interface connected to multiple front ports linked to different multi-position rear ports, or do we need to find a different way of modeling these devices?

**Reply by candlerb** [reply]
I can't answer that exactly, but I know you shouldn't be able to create an assertion error, so that in itself is a bug if you can show how to reproduce it from scratch (fill in the bug report form).

**Reply by mblue-fasttrack** [reply]
Will do, thank you for the assistance.

**Reply by kkthxbye-code** [reply]
Please make sure that your issue is separate from https://github.com/netbox-community/netbox/issues/11079 before submitting a new issue.

**Reply by mblue-fasttrack** [reply]
Issue is the same as #11079.
